server/src/main/java/org/opensearch/action/search/DeletePitAction.java

```diff
@@ -16,7 +16,7 @@
 public class DeletePitAction extends ActionType<DeletePitResponse> {
 
     public static final DeletePitAction INSTANCE = new DeletePitAction();
-    public static final String NAME = "indices:data/read/point_in_time/delete";
+    public static final String NAME = "cluster:admin/point_in_time/delete";
```

Comment:
@bharath-techie why it was moved to `cluster:admin`? It does not look right, the semantics of PIT should be kept under "indices:data/read", like `scroll` fe

Reply:
we initially went with "indices:read" , but then during testing, we found out that the user needs " all indices : read" permission to delete PITs. Same case with list all PITs.  This is because the API does not have indices as part of request.

We initially went with a scroll model but scroll does not have 'clear scroll' permission assigned to any role other than kibana_server. 

For more details :  please refer to https://github.com/opensearch-project/security/pull/2033

Reply:
It'll be great if you can review the overall security model as well @reta @Bukhtawar :
https://github.com/opensearch-project/OpenSearch/issues/3960 --- please review "Security model" section and the examples provided.

Reply:
okay i understand. but also on the flipside, thing to note is that 'pit search' and 'pit create' has indices as part of the request. So user only needs permission to that particular index.

But if user wants to delete pits or list all pits, user will need 'all indices read' permission. 

Are we okay with that ?

Reply:
"why all indices : read" permission is a problem?"

One issue is that -  just to access PIT APIs , we had to design a security role or action group which gives "indices - read all" permission.

There will be lack of granularity for users when assigning these roles/action groups

Reply:
Oh I see now, because delete / list just uses PIT ids (there is no index part), the are opening a possibility for user to see or delete PITs, created by other users. Shouldn't we extend the `security` plugin (fe [`IndexResolverReplacer`](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java) to resolve PITs to indices and apply the security checks?

Reply:
Delete all doesn't have any PIT IDs as part of reqeust : ( same with list all. 

Delete by id can use this yes.

Reply:
One approach is to call list all pits in rest layer to resolve pits and then call delete for "delete all". 
But. still "list all" cannot be contained with "indices:data/read" , it has to be part of "cluster:admin".

So don't see a way to contain all 3 apis as part of "indices:data/read".

Please suggest.

Reply:
> Delete all doesn't have any PIT IDs as part of reqeust : ( same with list all.

I think for this two, asking for having `all indices : read" permission` just makes sense

Reply:
@reta That might limit this operation to select users with elevated privileges, instead it would have been lot better if we could have listed/deleted all PITs accessible to the user with the current privileges f.e if user has read/delete access to lets say N indices, PITs pertaining to those N indices should be made accessible.

The way I see that might be implemented is have a `cluster:admin` privilege for the parent action which then resolves PIT to respective indices and filter PIT based on the scoped down index level privileges 

Short term I guess I am good with `all indices : read` from a security stand point since this would require elevated privileges but maybe from a use-case stand-point we might further relax this to still return a select list of PITs

Reply:
> Short term I guess I am good with `all indices : read` from a security stand point since this would require elevated privileges but maybe from a use-case stand-point we might further relax this to still return a select list of PITs

:+1: I believe the `security` plugin should be able to filter out PITs which current user is not able to see. The workarounds `cluster:admin` does not look appropriate here (to me): there is nothing related to cluster adminstration in PIT, this is purely search related feature.

Reply:
Based on this ^,  the flow will be something like 

```
1. User makes Delete all / List all API call
2. Rest Layer makes a call to Transport layer to get all PITs ('cluster:admin\getAllPits') -> 
     2.b. With this call, we will get initial PIT IDs in rest layer
3. Rest layer now passes the PIT IDs to Transport delete / get all ('indices:data/read/<action-name>')
    3.b. which can get decoded into indices and can be resolved in security plugin.
4. Only PIT IDs belonging to indices for which the user has permission will be returned back to user
```
So user will need following permissions:
```
cluster:admin/getallpits ----> this is for blanket pit read call which will retrieve PIT IDs to rest layer 
indices:data/read/point_in_time/read  & indices:data/read/point_in_time/delete --- this is the permission for actual APIs which will resolve index permissions
```

P.S. This model is where user don't need "indices read all" permission for list and delete all. We will perform both operations based on existing user permissions for indices.

"indices read all" sure can be a short term solution.

@reta what do you think about this ? not sure if we have any other way where we can avoid `cluster:admin/getallpits` - the initial internal call to retrieve all PITs

Reply:
thanks @Bukhtawar 

>  @reta are you suggesting we modify the security plugin to understand the PIT context and perform a filter based on this context?

The `security` plugin actually does that already, see please this guy https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java

> cluster:admin is just a proxy, we can call it something else,

Sadly it is not just a proxy - it leaks to admin client as well, for example. Using something else could be an option BUT the goal (from security pespective) is: the user should not be able to use PITs for indices she does not have access to. Right now filtering by action name means giving more priviledges. 

> If security plugin can already filter out PITs without dealing with the additional PIT context, the solution looks indeed promising. 

It does not do this now, but I think it is feasible to implement, `security` plugin knows the transport action it deals with, and it knows the user.

> @reta what do you think about this ? not sure if we have any other way where we can avoid cluster:admin/getallpits - the initial internal call to retrieve all PITs

I believe rest client should do nothing, the `security` plugin sits on server side and could  filter out the PITs belonging to indices for which the user has the permission.

Reply:
But for list all and delete all , there are no pit ids passed right - how will security plugin resolve the indices in that case - from security plugin even if we call server to get pit ids - still there will be a need of transport action which gets all the pit ids.

correct me if i'm wrong.

Reply:
That is correct, the way it could be implemented relies on the fact that indices are always passed when PITs are created (so we could reconstruct this link):
 - `security` plugin instuments List PITs action
 - once the PIT IDs are returned, `security` plugin fetches the indices assocaited with each PIT and filters out the ones user has no access to

Reply:
@reta @Bukhtawar 
I have made the changes in security plugin and opensearch to make PIT security model more granular as discussed which filters PITs based on underlying indices and associated permissions.

https://github.com/opensearch-project/security/pull/2053/files 

TransportNodesGetAllPITsAction - this will be the parent action ( not user / rest facing - transport only ) which will belong to 'cluster/admin' action type which initially gets all the active PITs from the cluster.

For Delete all , Delete list of PITs and Get All PITs APIs, security plugin will intercept the request and validate and permit the PITs based on underlying indices permission. 

We have a 'custom privilege evaluator' to handle above point in time requests.

Please refer to security PR https://github.com/opensearch-project/security/pull/2053/files  to know more about how each API filters the PITs.

Reply:
thanks @bharath-techie , this approach looks much better and cleaner (to me at least)

Reply:
@reta Here is my understanding, the current implementation does 
REST layer -> security plugin -> Transport(cluster:admin/point_in_time/read_from_nodes), where the security plugin evaluates what PITs should be accessible to the current user, with the caveat that security plugin doesn't have the original request's response listener as a result it requires to [block on a latch/interrupt](https://github.com/opensearch-project/security/pull/2053/files#diff-eb0f476c4343f7d278cdcc617b1de506134909cdccb93baf4effd7c914d0b043R321) ultimately blocking the request handler thread. While I am not sure how could we implement a callback without blocking threads with security plugin specifically.

Note :  the `cluster:admin` is still a proxy just that security plugin kicks in earlier before cluster:admin is hit with the earlier approach.

An alternative that was also discussed was somewhat similar(with few minor modification), which avoids the blocking nature of call made from the security plugin altogether.
```
1. User makes Delete all / List all API call
2. Rest Layer makes a call to Transport layer to get all PITs ('cluster:admin/getAllPits') -> 
     2.b. With this call, we will get initial PIT IDs 
3. Transport layer now passes the PIT IDs to Transport delete / get all ('indices:data/read/<action-name>')
    3.b. which can get decoded into indices and can be resolved in security plugin.
4. Only PIT IDs belonging to indices for which the user has permission will be returned back to user (or) deleted in case of delete all.
```

Thoughts?

Reply:
Please refer to this [commit](https://github.com/opensearch-project/security/pull/2053/commits/d5c482e123abcd10da1140043b5525527bc32c38#diff-54074a4a9d351028329ea8bccbab3cf6991374f8afcd39b28dffbba545589ad1R292) for latch usage , we've made changes to ensure that this is avoided , so latest changes don't have it @reta

Reply:
> Please refer to this [commit](https://github.com/opensearch-project/security/pull/2053/commits/d5c482e123abcd10da1140043b5525527bc32c38#diff-54074a4a9d351028329ea8bccbab3cf6991374f8afcd39b28dffbba545589ad1R292) for latch usage , we've made changes to ensure that this is avoided , so latest changes don't have it @reta

Yeah, already saw it, thanks @bharath-techie